corporate: Add required plan tier option for discounts.

laurynmm opened the pull request:
Adds the ability to set a required plan tier for a customer's default discount. Currently, only implements this support admin action for remote servers and realms.

**Notes**:
- When a discount is added to a customer, the flat $20 a month discount for the first year is no longer an option / available for that billing entity. We could add a note/warning in the support view about this if we want to highlight it. In general, the expectation is that these very specific discounts for particular organizations will be decided after a support conversation, so the loss of the first year, flat discount won't likely be an issue in practice.

---

**Screenshots and screen captures:**

<details>
<summary>Error message for setting required plan tier for discount with active plan</summary>

![Screenshot from 2024-01-10 17-58-39](https://github.com/zulip/zulip/assets/63245456/6ce2987d-e148-4cc8-947f-1d6824f15d98)
</details>

All screenshots below show the support view action with success message and then the two remote plan options with the current support state.

<details>
<summary>Set default discount to 90% - appears on all plan pages</summary>

![Screenshot from 2024-01-10 17-41-09](https://github.com/zulip/zulip/assets/63245456/48366210-cb62-457a-af03-a14227a93d9c)
</details>
<details>
<summary>Set required plan tier to Basic - 90% discount no longer on Business page</summary>

![Screenshot from 2024-01-10 17-41-26](https://github.com/zulip/zulip/assets/63245456/825c07cc-9a4d-4da9-a877-7826ff1a88e5)
</details>
<details>
<summary>Set required plan tier to Business - 90% discount no longer on Basic page</summary>

![Screenshot from 2024-01-10 17-41-42](https://github.com/zulip/zulip/assets/63245456/944d29de-a803-4b46-8487-7e737e81701c)
</details>
<details>
<summary>Set minimum licenses to 100 - appears on all plan pages</summary>

![Screenshot from 2024-01-10 17-42-06](https://github.com/zulip/zulip/assets/63245456/33bdac10-ca02-4c7c-a2ca-98e69c72b37f)
</details>
<details>
<summary>Set required plan tier to None - 90% discount on all plan pages & min licenses still 100</summary>

![Screenshot from 2024-01-10 17-42-29](https://github.com/zulip/zulip/assets/63245456/1ebd1279-4bc3-4ea3-8832-8b612cf6bfcc)
</summary>


---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

laurynmm commented on corporate/lib/stripe.py:
```diff
@@ -1205,9 +1205,8 @@ def setup_upgrade_payment_intent_and_charge(
-            plan_tier, billing_schedule, customer.default_discount
-        )
+        discount_for_plan = customer.get_discount_for_plan_tier(plan_tier)
+        price_per_license = get_price_per_license(plan_tier, billing_schedule, discount_for_plan)
```
We could also update `price_per_license` and `compute_plan_parameters` to take a customer object as a parameter (instead of the discount decimal) and get the discount as part of what happens in `price_per_license`.

laurynmm replied:
Due to the fact that we're likely using this new field for fixed rate plans, and will allow setting it when a billing entity has a current active plan, we won't want to change the behavior for `price_per_license` or `compute_plan_parameters` because for some cases we'll want to use the current plan discount (e.g. changing the billing schedule) and other times we'll want to use the customer default discount (and look at the plan tier).

timabbott replied:
Yeah, that sounds correct to me.

alya commented:
Screenshots look good to me!

laurynmm commented:
**Updates**:
- The new field on the Customer model is `required_plan_tier` as the plan is to use it both for plan percentage discounts and fixed rate plans.
- For percentage discounts, this is only checked when we are creating a new plan. For changes to existing plans that end users can make (e.g. change billing schedule), we want to use the discount that was already applied to the existing plan.
- Setting this via the support view does not require a customer default discount to already be set, so that we can use this for both discounts and flat rates.
- Added some notes to the support view to remind/alert admins how changing the plan tier or discount will impact existing plans. These are just formatted as italics in this iteration, but we can add some nicer formatting via CSS.

**Updated screenshots**:

There is no longer an error in the support view when a required plan tier is added to a billing entity with a current or scheduled plan.

<details>
<summary>Upgrade scheduled for Business - setting required plan tier to Basic - no change to scheduled plan or billing information for user</summary>
![Screenshot from 2024-01-11 19-44-45](https://github.com/zulip/zulip/assets/63245456/5fc9c9ea-c2ee-4d64-b37b-8a819a1a2665)
</details>
<details>
<summary>Changing the discount does update the scheduled upgrade to Business plan</summary>

![Screenshot from 2024-01-11 19-45-31](https://github.com/zulip/zulip/assets/63245456/46da91d2-3de3-4eaf-9e5f-1c89cd7c1fcd)
</details>

Series of screenshots for changing required plan tier / discount / min licenses for billing entity with no plan

<details>
<summary>Set required plan tier - no visible changes to upgrade billing pages</summary>

![Screenshot from 2024-01-11 19-48-44](https://github.com/zulip/zulip/assets/63245456/0d9406e2-e57b-4110-89df-74aff06f8993)
</details>
<details>
<summary>Set discount to Basic - appears on that plan billing page - flat-rate discount removed from Business</summary>

![Screenshot from 2024-01-11 19-48-59](https://github.com/zulip/zulip/assets/63245456/352f2d96-be47-4867-8c72-abab30fee3bf)
</details>
<details>
<summary>Add min licenses - on both pages</summary>

![Screenshot from 2024-01-11 19-49-11](https://github.com/zulip/zulip/assets/63245456/1ab1dd44-7602-4c39-96fe-5bb5569f2283)
</details>
<details>
<summary>Change required tier to Business - discount now only on that billing upgrade page</summary>

![Screenshot from 2024-01-11 19-49-28](https://github.com/zulip/zulip/assets/63245456/f07281fa-3aee-45ca-be91-528af5ed5f98)
</details>
<details>
<summary>Remove required plan tier - discount on both pages</summary>

![Screenshot from 2024-01-11 19-49-49](https://github.com/zulip/zulip/assets/63245456/9efd2e45-4b4f-48b4-8661-30b5dc4eaa6a)
</details>

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -1131,6 +1138,38 @@ def update_customer_minimum_licenses(self, new_minimum_license_count: int) -> st
+                "previous_required_plan_tier": previous_required_plan_tier,
+                "new_required_plan_tier": required_plan_tier,
+            },
+        )
```
We probably should be using the pattern we use in `RealmAuditLog.REALM_PROPERTY_CHANGED` rather than making up a new event log type for every field with its own unique format for every `Customer` field changed. Maybe a `CUSTOMER_PROPERTY_CHANGED` audit log event type?

laurynmm replied:
Added an initial prep commit for this change. There's now a `CUSTOMER_PROPERTY_CHANGED` and `CUSTOMER_PLAN_PROPERTY_CHANGED` for the audit logs. These replace `CUSTOMER_MINIMUM_LICENSES_CHANGED` and `CUSTOMER_PLAN_END_DATE_CHANGED` respectively.

I'm not sure if we've set a customer.minimum_license or updated any plan.end_date since these were added in the past week. We'd just need to update the extra_data on the log for the new format.

timabbott replied:
Yeah we can hand-fix this after this is merged, I think; very small numbers at least.

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -4014,6 +4083,16 @@ def get_upgrade_page_session_type_specific_context(
+        ]
+        if plan_tier in valid_tiers:
+            return True
+        return False
```
I'm not sure I love this function name, given that there are other valid plan tiers -- even if not valid to transition to in some cases.

laurynmm replied:
Changed from `check_valid_plan_tier` to `check_plan_tier_is_billable`. Another thought I had was `check_plan_tier_is_implemented`? It's mostly about whether the plan tier is appropriate for the billing entity (e.g. we don't want to accidentally set a Zulip Cloud tier on a self-hosted billing entity).

timabbott replied:
This is probably fine, though maybe worth a discussion in Zulip about further refactoring around tier definitions to make things nice and readable.

timabbott commented on templates/analytics/sponsorship_forms_support.html:
```diff
@@ -43,6 +44,22 @@
 
+<form method="POST" class="remote-form">
+    <b>Required plan tier for discounts and fixed prices</b>:<br />
+    <i>Updates will not change any pre-existing plans or scheduled upgrades.</i><br />
```
These hints are very helpful, thanks.

timabbott commented on templates/analytics/sponsorship_forms_support.html:
```diff
@@ -22,6 +22,7 @@
 
 <form method="POST" class="remote-form">
     <b>Discount</b>:<br />
+    <i>Updates will change pre-existing plans and scheduled upgrades.</i><br />
```
Do we want to clarify exactly how this works? E.g. I think it only will affect future charges -- existing prepaid will be fine.

laurynmm replied:
Sure. I added:

> **Discount percentage**:
> *Updates will change pre-existing plans and scheduled upgrades.*
> *Any prorated licenses for the current billing cycle will be billed at the updated discounted rate.*

So, prorated licenses added during a billing cycle are not billed until the next cycle's renewal payment. And they would all be at the new discounted rate in a future bill.

Feel free to edit that text though for more clarity.

timabbott commented:
lgtm aside from the comments above, thanks for all the work on this.

laurynmm commented:
**Updates**:
- Added a prep commit to simplify the audit log additions for changing a property on a Customer or CustomerPlan object.
- Addressed review feedback; see comments above.

<details>
<summary>Updated hints for discount and required plan tier support form</summary>

![Screenshot from 2024-01-12 20-51-19](https://github.com/zulip/zulip/assets/63245456/2a9ca9f2-19a4-45fc-b8bc-d39fd47ba85e)
</details>

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -1180,13 +1181,17 @@ def update_end_date_of_current_plan(self, end_date_string: str) -> str:
+                    extra_data={
+                        "old_value": old_end_date,
+                        "new_value": new_end_date,
+                        "property": "end_date",
```
We probably want to include the `plan_id` in these objects?

timabbott commented:
Made the tweak suggested in https://github.com/zulip/zulip/pull/28521#discussion_r1451029215 and marked this to merge once CI passes. Thanks for all the work on this @laurynmm!

timabbott commented:
This needs tests fixed for my tweak (only ran `test-backend corporate/` locally) and also a migration renumbering.

laurynmm commented:
**Update**:
- Renumbered migration
- Fixed test for audit log change
